### PR TITLE
Fix several bugs in compute_water_info and make the lichen boundry stable.

### PR DESCRIPTION
**Pull request by liqimai**
If a tile is already expanded, it should be added to the positions to water unconditionally. However, the current water info computation did not, and results in a strange behavior of lichen growth -- the unstable boundary. 

For example. consider the following case on a 1x4 map. There are two lichen strains, A and B. A and B initially occupy tiles 1 and 4. In the first turn, they both water, then they expanded to cell 2 and cell 3, respectively. Next turn, they both water again. However, they lost cell 2 and cell 3, respectively. This is because cell 2 is next to cell 3, which belongs to another strain, so it cannot be watered by A. Similarly, cell 3 cannot be watered by B. Finally, A lost cell 2, and B lost cell 3. If A and B keep watering, the two patterns will periodically appear, resulting in an unstable boundary, even if strains A and B belong to the same player. 

```
|A|    |A|    |A|    |A|
| | -> |A| -> | | -> |A| -> ...
| |    |B|    | |    |B|
|B|    |B|    |B|    |B|
```

In short, if a tile is already expanded and reachable from the factory, it should be added to the positions to water unconditionally, no matter whether its neighbors belong to other strains or not.

### CODE REVIEW

**Comment by liqimai** [comment]
Another two bugs in compute_water_info is fixed in second commit f0c233131b723c87095f3b0c7ba2b2c8ad4ad6bb:
1. line 44, we should compare `check_pos[0]` against `H`, and `check_pos[1]` aginst `W`.
2. Even If a `check_pos` is already seen, we still need to check its strain to decide whether the current `pos` is growable. Without this fix, a strain may wrongly expand to a position that is adjacent to other strains. However, according to the rule, "if a tile can be expanded to by two lichen strains, neither strain expands to there," so this fix is necessary.

**Comment by StoneT2000** [comment]
Thanks for the spot! I think this seems like a good change although I'll need to check a few things. Since this doesn't massively affect gameplay at the moment yet we might merge this in rather later due to high priority of other tasks.

**Comment by StoneT2000** [comment]
wait oops
